<a href="https://colab.research.google.com/github/masonnystrom/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
import pandas as pd
import numpy as np

In [0]:
column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

In [5]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names= column_headers)
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [6]:
vote_df = df.replace({"y":1, "n": 0, "?" : np.NaN},)
vote_df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [7]:
rep = vote_df[vote_df['party']=='republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [8]:
dem = vote_df[vote_df['party']=='democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

#### Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01


In [9]:
# let's find something polarizing. 
# examples: el salvador, anti-satellite, aid to contras, educaiton, south africa
dem.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [12]:
dem['anti-satellite-ban'].value_counts()

1.0    200
0.0     59
Name: anti-satellite-ban, dtype: int64

In [0]:
from scipy import stats

Null: Democrats and Reps support won't differ on the anti-satellite-ban.
Alt: Dems and Reps support for anti-satellite-ban will differ.

In [16]:
# t-test
stats.ttest_ind(dem['anti-satellite-ban'], rep['anti-satellite-ban'], nan_policy='omit')

Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)

Based on a t-stat of 12.5 and a p value of 8.521-31 we reject the null hypothesis that the support wouldn't differ in favor of the alternative hypothesis that support for the anti-stallite ban differs. 

### Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

In [10]:
rep.describe()
# wild guess, budget will be a hotly debated issue  

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

Null: Republicans and Democrats will not differ on their support for the budget bill

Alt: Republicans and democrats will differ on their support fo the budget bill. 

In [17]:
# run the t-test
stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

Based on a t-stat of 23 and a pvalue 2.07-77 (so small I don't want to type out that many zeros), we reject the null hypothesis in favor of the alternative that there is a difference in support between republicans and democrats on the budget bill. 

### Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Null: Reps and Dems support does not differ on ______ bill. 

Alt: Reps and Dems support differs on _____ bill. 

In [23]:
stats.ttest_ind(rep['aid-to-contras'], dem['aid-to-contras'], nan_policy='omit')
# reject null

Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

In [24]:
stats.ttest_ind(dem['el-salvador-aid'], rep['el-salvador-aid'], nan_policy='omit')
# reject Null

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

In [26]:
# maybe they both hate war?
stats.ttest_ind(dem['mx-missile'], rep['mx-missile'], nan_policy='omit')
# reject the null

Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)

In [28]:
# everybody loves suing people
stats.ttest_ind(dem['right-to-sue'], rep['right-to-sue'], nan_policy='omit')
# apparently not, reject the null

Ttest_indResult(statistic=-13.51064251060933, pvalue=1.2278581709672758e-34)

In [30]:
# education has to be bipartisan
stats.ttest_ind(rep['education'], dem['education'], nan_policy='omit')
# nope, reject the null

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

In [32]:
# religion
stats.ttest_ind(dem['religious-groups'], rep['religious-groups'], nan_policy='omit')
# reject the null

Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

In [34]:
stats.ttest_ind(dem['synfuels'], rep['synfuels'], nan_policy='omit')
# reject the null

Ttest_indResult(statistic=8.293603989407588, pvalue=1.5759322301054064e-15)

In [36]:
# crime
stats.ttest_ind(dem['crime'], rep['crime'], nan_policy='omit')

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.952342705606092e-47)

In [37]:
# duty free
stats.ttest_ind(dem['duty-free'], rep['duty-free'], nan_policy='omit')

Ttest_indResult(statistic=12.853146132542978, pvalue=5.997697174347365e-32)

### Success, the only bipartisan issue is water

In [35]:
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')
# Success, fail to reject the Null

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

Stretch goals:

Refactor your code into functions so it's easy to rerun with arbitrary variables


Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
def voting(group1[], group2[]):
  return stats.ttest_ind(group1[], group2[], nan_policy='omit')

In [41]:
voting(dem['water-project'], rep['water-project'])

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)